### PR TITLE
[DNM] drivers: flash: soc_flash_nrf: add support for nrf9160

### PR DESCRIPTION
**Pull request by lemrey**
This PR is intended to give visibility to the workaround to necessary to use Zephyr's flash driver on nRF9160.

FYI: @gbakke @osaether

### CODE REVIEW

File: drivers/flash/soc_flash_nrf.c [hunk 1]
```diff
@@ -77,6 +78,17 @@ static struct k_sem sem_lock;
 #define SYNC_UNLOCK()
 #endif
 
+#if defined(CONFIG_SOC_SERIES_NRF51X)
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_nrf.c]: you resign form ficr usage, maybe it is better.

File: drivers/flash/soc_flash_nrf.c [hunk 2]
```diff
@@ -127,8 +139,33 @@ static void nvmc_wait_ready(void)
 	}
 }
 
-static int flash_nrf_read(struct device *dev, off_t addr,
-			    void *data, size_t len)
+enum nvmc_config {
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_nrf.c]: by default enums are int size, but it possible to switch to other size.
Fro the other hand nice to use this in func. declaration.
Maybe no worth to be so carefully.

File: drivers/flash/soc_flash_nrf.c [hunk 3]
```diff
@@ -217,12 +254,7 @@ static int flash_nrf_write_protection(struct device *dev, bool enable)
 {
 	SYNC_LOCK();
 
-	if (enable) {
-		NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Ren << NVMC_CONFIG_WEN_Pos;
-	} else {
-		NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Wen << NVMC_CONFIG_WEN_Pos;
-	}
-	nvmc_wait_ready();
+	nvmc_config_set(enable ? REN : WEN);
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_nrf.c]: `(void)` before will suppress warning

File: drivers/flash/soc_flash_nrf.c [hunk 4]
```diff
@@ -432,8 +464,7 @@ static int write_in_timeslice(off_t addr, const void *data, size_t len)
 
 static int erase_op(void *context)
 {
-	u32_t prev_nvmc_cfg = NRF_NVMC->CONFIG;
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_nrf.c]: variable should be kept

File: drivers/flash/soc_flash_nrf.c [hunk 5]
```diff
@@ -483,8 +517,7 @@ static int erase_op(void *context)
 
 	} while (e_ctx->len > 0);
 
-	NRF_NVMC->CONFIG = prev_nvmc_cfg;
-	nvmc_wait_ready();
+	nvmc_config_set(REN);
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_nrf.c]: `nvmc_config_set(prev_nvmc_cfg);`

File: drivers/flash/soc_flash_nrf.c [hunk 6]
```diff
@@ -446,9 +477,8 @@ static int erase_op(void *context)
 	}
 #endif /* CONFIG_SOC_FLASH_NRF_RADIO_SYNC */
 
-	/* Erase uses a specific configuration register */
-	NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Een << NVMC_CONFIG_WEN_Pos;
-	nvmc_wait_ready();
+	/* Enable erase operation */
+	nvmc_config_set(EEN);
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_nrf.c]: prev_nvms_cfg = nvmc_config_set(EEN);

File: drivers/flash/soc_flash_nrf.c [hunk 7]
```diff
@@ -513,6 +546,9 @@ static int write_op(void *context)
 	}
 #endif /* CONFIG_SOC_FLASH_NRF_RADIO_SYNC */
 
+	/* Enable write operation */
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_nrf.c]: f. flash_nrf_write_protection() is for enable write explicite. Pleas remove this form here - its breaks API compatibility.

**Comment by IDEMIA-Pessac** [comment]
If I may jump in, there are some issues with the proposed patch.
The flash write and erase features won't be working with 
`CONFIG_ARM_SECURE_FIRMWARE=y`

I had to slightly modify the usage of `NRF_NVMC->CONFIG` like this : 

```
enum nvmc_config {
#if defined(CONFIG_SOC_NRF9160) && !defined(CONFIG_ARM_SECURE_FIRMWARE)
	REN = NVMC_CONFIGNS_WEN_Ren << NVMC_CONFIGNS_WEN_Pos,
	WEN = NVMC_CONFIGNS_WEN_Wen << NVMC_CONFIGNS_WEN_Pos,
	EEN = NVMC_CONFIGNS_WEN_Een << NVMC_CONFIGNS_WEN_Pos,
#else
	REN = NVMC_CONFIG_WEN_Ren << NVMC_CONFIG_WEN_Pos,
	WEN = NVMC_CONFIG_WEN_Wen << NVMC_CONFIG_WEN_Pos,
	EEN = NVMC_CONFIG_WEN_Een << NVMC_CONFIG_WEN_Pos,
#endif
};

static u32_t nvmc_config_set(enum nvmc_config mode)
{
#if defined(CONFIG_SOC_NRF9160) && !defined(CONFIG_ARM_SECURE_FIRMWARE)
	const u32_t cfg = NRF_NVMC->CONFIGNS;
	NRF_NVMC->CONFIGNS = mode;
#else
	const u32_t cfg = NRF_NVMC->CONFIG;
	NRF_NVMC->CONFIG = mode;
#endif
	nvmc_wait_ready();
	return cfg;
}
```

**Comment by nvlsianpu** [comment]
replaced by #118

**Pull request closed without merging**